lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,402 @@
+ * cache to each bundle host object AFTER it is instantiated because running the
+ * addon constructor ultimately causes Addon class `setupRegistry` code to
+ * run which instantiates child addons, which need the cache to already be
+ * in place for the parent bundle host.
+ * We handle this by providing a global cache that exists independent of the
+ * bundle host objects. That is this object.
+ *
+ * There are a number of "behaviors" being implemented by this object and
+ * its contents. They are:
+ * (1) Any addon that is a lazy engine has only a single real instance per
+ * project - all other references to the lazy engine are to be proxies. These
+ * lazy engines are compared by name, not by packageInfo.realPath.
+ * (2) Any addon that is not a lazy engine, there is only a single real instance
+ * of the addon per "bundle host" (i.e. lazy engine or project).
+ * (3) An optimization - any addon that is in a lazy engine but that is also
+ * in the project (outside of all lazy engines) - the single instance is the
+ * one in the project. All other instances (in any lazy engine) are proxies.
+ * NOTE: the optimization is only enabled if the environment variable that controls that
+ * ember-engine deduplication (process.env.EMBER_ENGINES_ADDON_DEDUPE) is set
+ * to a truthy value.
+ *
+ * Some implementation details given the above desired behavior:
+ * (1) There are actually 2 types of caches in this object:
+ * (1a) the first is keyed by lazy engine name. Note: the real instance of
+ * the lazy engine is created as it is encountered while traversing the addon
+ * tree (just like any real addon instance), it's just also referenced
+ * in this cache so we can create proxies to this single instance from
+ * anywhere in the addon tree.
+ * (2) The cache (one item per lazy engine + project) of real addon instances
+ * found while traversing the dependency tree.
+ *
+ * @public
+ * @class PerBundleAddonCache {
+ */
+class PerBundleAddonCache {
+  constructor(project) {
+    this.project = project;
+
+    // cache of the addon packageInfos for addons that are descendants of the project
+    // (excluding going into lazy engines). Lazy engines that refer to any of these
+    // will defer to the project-level one.
+    this.bundledProjectAddonPkgInfos = this._setupBundledProjectAddonPackageInfos();
+
+    // The cache of bundle-host instances and their individual addon caches.
+    // The cache is keyed by name (since there is only a single instance of any given
+    // engine name, even if 2 engine addons have the same engine name).
+    // To allow the project in the cache, it will use the empty string for a name,
+    // to not collide with any possible engine name.
+    // Each cache entry consists of the bundle host instance (Project or Lazy engine)
+    // and an addon instance cache to bundle with that bundle host.
+    this.bundleHostCache = this._setupBundleHostCache();
```

Comment:
do we want to set these caches up eagerly? if nobody enables `allowCachingPerBundle` for any addon, then ppl will incur this cost and these caches will go unused

Reply:
I get your point. Just FYI the cost of setup is << 1 second. But I'll look more into the complexity of what has to be checked where to make sure the caches are or are not created. Remember that the addon instance cache for a given lazy engine MUST exist BEFORE the lazy engine itself has finished its constructor, because of that g*&(&mn `setupRegistry` call to create the children before the parent is done. It cannot wait until there is an instance of a child that wants to be put into the addon cache for that bundle host, as we don't have access to the inside of the bundle host's constructor function.

Reply:
Ultimately decided to leave setting up the caches as is for now, given the small cost.